File: internal/study/survey_statistics.go
package study

import (
	"math"
	"slices"
	"strconv"
	"strings"
	"time"
)

type surveyRate struct {
	Events int `json:"events"`
	Total  int `json:"total"`
}
type surveyAssociation struct {
	Selected   surveyRate `json:"selected"`
	Unselected surveyRate `json:"unselected"`
	Phi        *float64   `json:"phi"`
}

func association(a, b, c, d int) surveyAssociation {
	result := surveyAssociation{Selected: surveyRate{a, a + b}, Unselected: surveyRate{c, c + d}}
	denominator := math.Sqrt(float64(a+b) * float64(c+d) * float64(a+c) * float64(b+d))
	if denominator > 0 {
		phi := (float64(a)*float64(d) - float64(b)*float64(c)) / denominator
		result.Phi = &phi
	}
	return result
}

type surveyCount struct {
	Label string `json:"label"`
	Count int    `json:"count"`
}
type surveyGroup struct {
	Total int           `json:"total"`
	Rows  []surveyCount `json:"rows"`
}
type surveyFactor struct {
	ID          string                 `json:"id"`
	Title       string                 `json:"title"`
	Description string                 `json:"description"`
	Multiple    bool                   `json:"multiple"`
	Eligibility string                 `json:"eligibility"`
	Groups      map[string]surveyGroup `json:"groups"`
}
type surveyAssociationRow struct {
	Label    string            `json:"label"`
	Degraded surveyAssociation `json:"degraded"`
	Banned   surveyAssociation `json:"banned"`
	Limited  surveyAssociation `json:"limited"`
}
type surveyAssociationGroup struct {
	ID    string                 `json:"id"`
	Title string                 `json:"title"`
	Scope string                 `json:"scope"`
	Total int                    `json:"total"`
	Rows  []surveyAssociationRow `json:"rows"`
}
type surveyStatusCount struct {
	Label string `json:"label"`
	Count int    `json:"count"`
	Tone  string `json:"tone"`
}
type surveyUsagePattern struct {
	Total  int         `json:"total"`
	Levels [24]float64 `json:"levels"`
}
type surveyStatistics struct {
	Total        int                      `json:"total"`
	Degraded     int                      `json:"degraded"`
	Banned       int                      `json:"banned"`
	Limited      int                      `json:"limited"`
	Statuses     []surveyStatusCount      `json:"statuses"`
	Normal       int                      `json:"normal"`
	Both         int                      `json:"both"`
	Factors      []surveyFactor           `json:"factors"`
	Associations []surveyAssociationGroup `json:"associations"`
	UsagePattern surveyUsagePattern       `json:"usagePattern"`
	Range        surveyStatisticsRange    `json:"range"`
}

type surveyStatisticsRange struct {
	FirstSubmissionID uint64     `json:"firstSubmissionId"`
	LastSubmissionID  uint64     `json:"lastSubmissionId"`
	FirstSubmittedAt  *time.Time `json:"firstSubmittedAt"`
	LastSubmittedAt   *time.Time `json:"lastSubmittedAt"`
	UnknownTimeCount  int        `json:"unknownTimeCount"`
	ComputedAt        time.Time  `json:"computedAt"`
}

func (q SurveySubmission) normalized() map[string][]string {
	result := make(map[string][]string, len(q.Answers)+8)
	for key, value := range q.Answers {
		result[key] = value
	}
	add := func(key, value string) { result[key] = []string{value} }
	if q.IPRisk != nil {
		index := 5
		for i, limit := range []int{15, 25, 40, 50, 70} {
			if *q.IPRisk < limit {
				index = i
				break
			}
		}
		add("ipRisk", surveyCatalog.Choices["ipRisk"][index])
	}
	for _, key := range []string{"country", "exitCountry"} {
		code := q.Details[key]
		if code == "" {
			continue
		}
		name := map[string]string{"PH": "菲律宾", "JP": "日本", "US": "美国", "BO": "玻利维亚", "unknown": "我不知道"}[code]
		if name == "" {
			name = "其他国家或地区"
		}
		add(key, name)
	}
	if raw := q.Details["duration"]; raw != "" {
		days, _ := strconv.ParseFloat(raw, 64)
		days *= map[string]float64{"天": 1, "小时": 1.0 / 24, "星期": 7, "月": 30, "年": 365}[q.Details["durationUnit"]]
		index := 5
		for i, limit := range []float64{1, 7, 30, 90, 365} {
			if days < limit {
				index = i
				break
			}
		}
		add("duration", surveyCatalog.Choices["duration"][index])
	}
	if raw := q.Details["people"]; raw != "" {
		number, _ := strconv.Atoi(raw)
		index := 4
		for i, limit := range []int{2, 5, 10, 20} {
			if number <= limit {
				index = i
				break
			}
		}
		add("people", surveyCatalog.Choices["people"][index])
	}
	if raw := q.Details["concurrency"]; raw != "" {
		index := 6
		if raw != "unknown" {
			number, _ := strconv.Atoi(raw)
			index = 5
			for i, limit := range []int{0, 1, 3, 5, 10} {
				if number <= limit {
					index = i
					break
				}
			}
		}
		add("concurrency", surveyCatalog.Choices["concurrency"][index])
	}
	for key, value := range q.Details {
		if strings.HasPrefix(key, "toolMode:") && !slices.Contains(result["thirdMode"], value) {
			result["thirdMode"] = append(result["thirdMode"], value)
		}
	}
	return result
}

File: web/src/data/surveyStatistics.ts
import type { StatusCounts, SurveyFactorSummary, SurveyOutcome, SurveySummary } from "./surveyApi";
import type { BinaryAssociation } from "./surveyAssociation";

export const surveyOutcomes: { id: SurveyOutcome; label: string; bit: number }[] = [
  { id: "degraded", label: "降智", bit: 1 },
  { id: "banned", label: "封号", bit: 2 },
  { id: "limited", label: "风控（限流）", bit: 4 },
];
export interface FactorDistribution {
  id: string;
  title: string;
  description: string;
  multiple: boolean;
  eligibility: string;
  groups: Record<SurveyOutcome, { total: number; rows: { label: string; count: number }[] }>;
}
export interface AssociationGroup {
  id: string;
  title: string;
  scope: string;
  total: number;
  rows: { label: string; association: BinaryAssociation }[];
}
export interface SurveyStatistics {
  total: number;
  degraded: number;
  banned: number;
  limited: number;
  normal: number;
  statuses: { label: string; count: number; tone: string }[];
  factors: FactorDistribution[];
  usagePattern: { total: number; levels: number[] };
  range: SurveySummary["range"];
}

export function countStatuses(counts: StatusCounts, mask?: number): number {
  let total = 0;
  for (let state = 0; state < 8; state++) {
    if (mask === undefined || (state & mask) !== 0) total += counts[state]!;
  }
  return total;
}

export function effectiveOutcomeMask(
  selected: readonly SurveyOutcome[],
  lastSelected: SurveyOutcome,
): number {
  let mask = 0;
  for (const item of surveyOutcomes) if (selected.includes(item.id)) mask |= item.bit;
  return mask || surveyOutcomes.find((item) => item.id === lastSelected)!.bit;
}

export function calculateAssociation(
  selected: StatusCounts,
  applicable: StatusCounts,
  mask: number,
): BinaryAssociation {
  const selectedTotal = countStatuses(selected);
  const total = countStatuses(applicable);
  const a = countStatuses(selected, mask);
  const b = selectedTotal - a;
  const c = countStatuses(applicable, mask) - a;
  const d = total - a - b - c;
  const denominator = Math.sqrt((a + b) * (c + d) * (a + c) * (b + d));
  return {
    selected: { events: a, total: selectedTotal },
    unselected: { events: c, total: total - selectedTotal },
    phi: denominator > 0 ? (a * d - b * c) / denominator : null,
  };
}

export function calculateAssociations(
  factors: SurveyFactorSummary[],
  mask: number,
): AssociationGroup[] {
  return factors
    .filter((factor) => factor.comparable)
    .map((factor) => ({
      id: factor.id,
      title: factor.title,
      scope: factor.associationScope,
      total: countStatuses(factor.applicable),
      rows: factor.options.map((option) => ({
        label: option.label,
        association: calculateAssociation(option.counts, factor.applicable, mask),
      })),
    }));
}

const statusLabels = [
  "正常",
  "降智",
  "封号",
  "降智、封号",
  "风控（限流）",
  "降智、风控（限流）",
  "封号、风控（限流）",
  "降智、封号、风控（限流）",
];
const statusTones = ["mint", "violet", "peach", "peach", "sun", "sun", "sun", "sun"];

export function calculateSurveyStatistics(summary: SurveySummary): SurveyStatistics {
  const factors = summary.factors.map((factor): FactorDistribution => {
    const groups = {} as FactorDistribution["groups"];
    for (const outcome of surveyOutcomes) {
      groups[outcome.id] = {
        total: countStatuses(factor.applicable, outcome.bit),
        rows: factor.options.map((option) => ({
          label: option.label,
          count: countStatuses(option.counts, outcome.bit),
        })),
      };
    }
    return {
      id: factor.id,
      title: factor.title,
      description: factor.description,
      multiple: factor.multiple,
      eligibility: factor.distributionScope,
      groups,
    };
  });
  return {
    total: countStatuses(summary.statuses),
    degraded: countStatuses(summary.statuses, 1),
    banned: countStatuses(summary.statuses, 2),
    limited: countStatuses(summary.statuses, 4),
    normal: summary.statuses[0],
    statuses: summary.statuses.map((count, mask) => ({
      label: statusLabels[mask]!,
      tone: statusTones[mask]!,
      count,
    })),
    factors,
    usagePattern: {
      total: summary.usagePattern.total,
      levels: summary.usagePattern.sums.map((sum) =>
        summary.usagePattern.total ? sum / summary.usagePattern.total : 0,
      ),
    },
    range: summary.range,
  };
}
